bazel/astore/defs.bzl

```diff
@@ -60,3 +60,47 @@ astore_upload = rule(
 With this rule, you can easily upload the output of a build rule
 to an artifact store.""",
 )
+
+def _astore_download(ctx):
```

Comment:
ok, almost :) so... when you invoke bazel to do something, you can generally do three things: `run` a target, or `build` the target, or `test` the target. Forget about test.

* `run` -> performs an action, does something (for example, deploys a binary to production).
* `build` -> creates something to be used by the rest of the build.

Now... the `_astore_upload` run I created is a `run` kind of rule. The idea is that the user builds a bunch of stuff, tests it out, and then when he's happy with the build, will finally deploy to astore. So the rule was structured so that the user would call `bazel build whatever` and then `bazel run :upload` to send to astore.

So, what the `_astore_upload` rule does is *create something executable* to be *run later*, when the *user asks for it* by separately invoking `bazel run ...`.

`_astore_download` needs to be a `build` rule: it needs to immediately `create` a file to be used by the rest of the build process. `DefaultInfo` should not return an `executable`, but a set of `files` to use.

This is what I recommend you do:
* instead of using `ctx.actions.expand_template`, use `ctx.actions.run` or `ctx.actions.run_shell` (https://docs.bazel.build/versions/master/skylark/lib/actions.html#run). Ditch the astore_download.sh file - it's a simple enough command, that you don't need a separate shell script.
* in `actions.run`, specify your `ctx.executable._astore_client` as a `tool` the rule needs, rather than as an input. So you can ditch the `inputs` variable as well.
* create an `outputs` array, listing the outputs the rules create (use `ctx.action.declare_file` on the `ctx.attr.file` to create a file object). Pass this `outputs` array as the argument `outputs` of `run` or `run_shell`.

And finally, super important: change the DefaultInfo. This rule is no longer creating an executable someone can run, with a set of inputs it has to process. Rather it is just creating a set of output files.

Should be as simple as: `return [DefaultInfo(files = outputs)]`

Reply:
ah! yes. Imagine you have a function that expects a python set, you are passing in a list, that's the error.

To convert a list to a depset, based on the example here:
https://docs.bazel.build/versions/master/skylark/lib/globals.html#depset

you probably just need to do

```
return [DefaultInfo(files = depset(outputs))]

Reply:
the type conversion... in the _kernel_module macro I have:
```
    output = ctx.actions.declare_file(rename)
...
    return DefaultInfo(files = depset([output]))
```

which should just work? I have seen the use of list comprehension otherwise, eg, files = despet([ ... for f in ...])